pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -376,6 +400,211 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+
+        for goArch in util.goArches:
+            if goArch == "multi":
+                continue
+            # From the newly built release, extract the client tools into the workspace following the directory structure
+            # we expect to publish to mirror
+            CLIENT_MIRROR_DIR = f"{RELEASE_MIRROR_DIR}/{goArch}"
+            os.makedirs(CLIENT_MIRROR_DIR)
+            # extract release clients tools
+            extract_release_client_tools(f"{constants.QUAY_URL}:{from_release_tag}", f"--to={CLIENT_MIRROR_DIR}", goArch)
+            # create symlink for clients
+            self.create_symlink(CLIENT_MIRROR_DIR, True, True)
+
+        # Create a master sha256sum.txt including the sha256sum.txt files from all subarches
+        # This is the file we will sign -- trust is transitive to the subarches
+        os.chdir(RELEASE_MIRROR_DIR)
+        for dir in os.listdir(RELEASE_MIRROR_DIR):
+            if not os.path.isdir(os.path.join(RELEASE_MIRROR_DIR, dir)):
+                continue
+            for root, dirs, files in os.walk(os.path.join(RELEASE_MIRROR_DIR, dir)):
+                if "sha256sum.txt" not in files:
+                    continue
+                with open(os.path.join(root, "sha256sum.txt"), "rb") as f:
+                    shasum = hashlib.sha256(f.read()).hexdigest()
+                with open(f"{RELEASE_MIRROR_DIR}/sha256sum.txt", 'a') as f:  # write shasum to sha256sum.txt
+                    f.write(f"{shasum} {dir}/sha256sum.txt")
+
+        # Publish the clients to our S3 bucket.
+        await exectools.cmd_assert_async(f"aws s3 sync --no-progress --exact-timestamps {BASE_TO_MIRROR_DIR}/ s3://art-srv-enterprise/pub/openshift-v4/", stdout=sys.stderr)
+
+    def create_symlink(self, path_to_dir, print_tree, print_file):
+        # External consumers want a link they can rely on.. e.g. .../latest/openshift-client-linux.tgz .
+        # So whatever we extract, remove the version specific info and make a symlink with that name.
+        for f in os.listdir(path_to_dir):
+            if f.endswith(('.tar.gz', '.bz', '.zip', '.tgz')):
+                # Is this already a link?
+                if os.path.islink(f):
+                    continue
+                # example file names:
+                #  - openshift-client-linux-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-client-mac-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-install-mac-4.3.0-0.nightly-2019-12-06-161135.tar.gz
+                #  - openshift-client-linux-4.1.9.tar.gz
+                #  - openshift-install-mac-4.3.0-0.nightly-s390x-2020-01-06-081137.tar.gz
+                #  ...
+                # So, match, and store in a group, any character up to the point we find -DIGIT. Ignore everything else
+                # until we match (and store in a group) one of the valid file extensions.
+                match = re.match(r'^([^-]+)((-[^0-9][^-]+)+)-[0-9].*(tar.gz|tgz|bz|zip)$', f)
```

Comment:
It would be handy to have unit tests to test we're getting the name that we expect.
Would suggest either extracting it out in a separate func, and then testing that

Reply:
Unresolve. We need bits of this PR covered in unit tests.

Reply:
ut added